Pcs different sizes

wreise opened the pull request:
#### Reference Issues/PRs
Continuation of #197, hence the copied description. Fixes #144 and fixes #299.

#### What does this implement/fix? Explain your changes.
All the simplex and grid methods should now accept point clouds.

#### Any other comments?
1. Verify the behavior of `check_list_of_arrays`, depending on the value of `copy`. Is it consistent with the behavior of `check_array` ?

wreise committed:
Add a check_list_of_arrays fct in utils/validation

wreise committed:
Change the checks in the homology classes

wreise committed:
Discard check on arrays and treat everything as lists in utils/validations:check_list_of_arrays, modify the point-cloud class accordingly

wreise committed:
Adapt the tests for pcs

wreise committed:
Correct the indentation on the for loop

wreise committed:
Ensure that all points clouds are 2d tensors

wreise committed:
Introduce the check_list_of_images function

wreise committed:
Use check_point_clouds in the point-cloud filtrations. Add the allow_nd param to check_array for images

wreise committed:
Fix lkinting

wreise committed:
Change the size of inputs in preprocessing

wreise committed:
Make the outpu of the image2pctransformer variable in size,  and adjust the tests (modify the expected values)

wreise committed:
Fix linting in gttda.images and the docstring from imageToPointCloud

wreise committed:
Document the validation functions, lint test preprocessing and make ImageToPointCloud pass the check_is:_fitted test

wreise committed:
Work on docstring for check images  and point cloud functions

wreise committed:
Merge branch 'master' into pcs_different_sizes

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+
+
+def check_list_of_images(X, **kwargs):
+    """"Check a list of arrays representing images, by integrating
```
Looks like 4 quotation marks here.

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+def check_list_of_images(X, **kwargs):
+    """"Check a list of arrays representing images, by integrating
+    through the input one by one. To pass a test with
+    :param:`kwargs`=``None``, all images x in X must satisfy:
```
AFAIK, :param: does not render the way we'd think for some reason. It is never used in the library.

ulupo replied:
Change to "when `kwargs` is ``None``" (single ticks for kwargs, double for None)

ulupo replied:
Also change X for `X` (single ticks), x for ``x`` (double ticks). The philosophy here is that single ticks are reserved to parameter names, and double for everything else (of course attributes, methods, functions, and classes are treated differently, using :attr:, :meth:, :func: and :class: respectively.

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+    """"Check a list of arrays representing images, by integrating
+    through the input one by one. To pass a test with
+    :param:`kwargs`=``None``, all images x in X must satisfy:
+        - x.ndim >= 2,
```
Surround with double ticks here and below.

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+
+    Parameters
+    ----------
+    X : list of ndarray,
```
Remove comma

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+    Parameters
+    ----------
+    X : list of ndarray,
+        Each entry of X corresponds to an image.
```
Single ticks around X

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+    X : list of ndarray,
+        Each entry of X corresponds to an image.
+
+    kwargs: dict or None, optional, default: ``None``
```
Add space between kwargs and colon

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+
+    Returns
+    -------
+    X : list of ndarray, as modified by check_array
```
Remove comma and place everything following the comma in a description below

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+
+
+def check_list_of_point_clouds(X, **kwargs):
+    """"Check a list of arrays representing point clouds, by integrating
```
Implement same (where applicable) dosctring fixes as for `check_list_of_images`.

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+    return check_list_of_arrays(X, **kwargs_default)
+
+
+def check_list_of_arrays(X, force_same_dim=True, force_same_n_axis=True,
```
Implement same (where applicable) dosctring fixes as for `check_list_of_images`

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+
+    Parameters
+    ----------
+    X : list(ndarray), such that `X[i].ndim==2` (n_points, n_dimensions),
```
"list of ndarray"

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+    if force_same_dim:
+        shapes = [x.shape for x in X]
+        if not(all([shapes[0] == s for s in shapes])):
+            raise ValueError("The arrays in X do not have the same dimensions"
```
Since the move to Python 3.6, we can and should use f-strings

ulupo replied:
In fact I'll open a separate PR/issue about refactoring this.

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+    """
+
+    # if restrictions on the dimensions of the input are imposed
+    if force_same_dim:
```
Since you are checking for the same *shape*, this variable should perhaps be called `force_same_shape`

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+                             "({}), while they should.".format(shapes))
+    # if the number of dimensions can vary
+    if force_same_n_axis:
+        n_axis = [len(x.shape) for x in X]
```
Why not simply use `x.ndim`?

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+            raise ValueError("The arrays in X do not have the same dimensions"
+                             "({}), while they should.".format(shapes))
+    # if the number of dimensions can vary
+    if force_same_n_axis:
```
If you end up using `ndim`, it would make more sense for this to be called `force_same_ndim`

ulupo replied:
`force_same_dim`(AKA `force_same_shape`) supersedes the  calculations here so I'd suggest restructuring the code so that no unnecessary checks are made.

wreise committed:
Fix docstring issues

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,117 @@ def validate_metric_params(metric, metric_params):
+    for id_x, x in enumerate(X):
+        try:
+            X[id_x] = check_array(x, **kwargs)
+            results.append(True)
```
The use of `results` seems unnecessary: I think this part of the code could be rewritten using the `for ... else` paradigm, I can try making a proposal if OK with you.

ulupo replied:
Scrap my comment about `for ... else`, but I do think that `results` should be replaced by a boolean flag.

wreise committed:
Rename force_same_dim to force_same_shape

ulupo commented on gtda/utils/validation.py:
```diff
@@ -174,3 +176,120 @@ def validate_metric_params(metric, metric_params):
+    """Check a list of arrays representing images, by integrating
+    through the input one by one. To pass a test when `kwargs` is ``None``,
+    all images ``x`` in `X` must satisfy:
+        - ``x.ndim`` >= 2,
```
I think this will look bad without including ">= 2" in the double ticks.

wreise committed:
Remove list of results, andreplaceitwith aflag parameter, put ToDo

wreise committed:
Change force_same_n_axis to force_same_ndim and fix a docstring

ulupo committed:
Docstring fixes

wreise committed:
Merge with master

wreise committed:
Remove  redundant lines for test results definitions for ImageToPointCloud

wreise committed:
Modularizethe checks with lambda functions. Modify the tests in test_simplicial (tests for ok and invalid shapes respectively)

wreise committed:
Add a docstring in check_dimensions

wreise committed:
Adapt the docstrings of check_list_of_point_clouds and check_list_of_images, to match the logic agreed upon

wreise committed:
Add falling back to check_array if the input is really an array

wreise committed:
Fix linting invalidation

wreise committed:
Add a fail on small dimensional-images

wreise committed:
Resolve merge conflicts

wreise committed:
Merge  master into pcs_different_sizes

wreise committed:
Merge master(2)

ulupo commented on gtda/utils/validation.py:
```diff
@@ -154,3 +156,151 @@ def validate_params(parameters, references, exclude=None):
+
+
+def check_list_of_images(X, **kwargs):
+    """Check a list of arrays representing images, by integrating
```
Should "integrating" be "iterating"?

ulupo commented on gtda/utils/validation.py:
```diff
@@ -154,3 +156,151 @@ def validate_params(parameters, references, exclude=None):
+
+def check_list_of_images(X, **kwargs):
+    """Check a list of arrays representing images, by integrating
+    through the input one by one. To pass a test when `kwargs` is ``None``,
```
"`kwargs` is empty"?

ulupo replied:
(And same for other validation functions in this PR.)

wreise replied:
Should be ok now

ulupo commented on gtda/utils/validation.py:
```diff
@@ -154,3 +156,151 @@ def validate_params(parameters, references, exclude=None):
+    X : list of ndarray
+        Each entry of `X` corresponds to an image.
+
+    kwargs : dict or None, optional, default: ``None``
```
The default is the empty dictionary?

ulupo replied:
(And same for other validation functions in this PR.)

wreise replied:
I followed your suggestion. is it resolved?

wreise committed:
Change docstrings on kwargs in utils.validation

wreise committed:
Add code formatting in docstrings in utils.validation

wreise committed:
Add sklearn estimator checks known failures

ulupo commented on gtda/utils/validation.py:
```diff
@@ -154,3 +156,151 @@ def validate_params(parameters, references, exclude=None):
+    X : list of ndarray
+        Each entry of `X` corresponds to an image.
+
+    kwargs : dict, optional, default: ``{}``
```
I'm sorry. I should have thought it better before suggesting ``{}``. In reality I don't think this is good either as the user would not actually pass a dictionary here, so it is misleading. I think we should adopt an approach as in https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html, i.e.
```
**kwargs
    Keyword arguments accepted by
    :func:`~gtda.utils.validation.check_list_of_arrays`.
```

wreise committed:
Change the kwargs docuemntation in validation

wreise committed:
Merge with upstream master

ulupo committed:
Merge branch 'master' into pcs_different_sizes

Commit:
Merge branch 'master' into pcs_different_sizes

Commit:
Rename functions

check_list_of_point_clouds -> check_point_clouds
check_list_of_images -> check_images

Commit:
Fixes in ImageToPointCloud

Commit:
Add mapper and utils to global __init__

Commit:
Make check_shapes a list of pairs, fix validation docstrings, revise use of check_array throughout

Commit:
Add check_point_clouds to utils __init__ and docs

Commit:
Remove check_images and additions to test_common

These are valuable and thank you @wreise, but in the interest of pushing at least check_point_clouds, I am opting for a more ad-hoc implementation of the latter and for the temporary removal of check_images. As we also did not have much time to discuss the additions to test_common, I will also revert that file to the state on master for now. It can be reintroduced in a later PR together with check_images

Commit:
Linting

Commit:
Refactor ImageToPointCloud to return lists of point clouds

Commit:
Refactor check_point_clouds making it more ad-hoc

Commit:
Fix some incorrect links to GH pages

Commit:
Delete test expecting ValueError on point clouds in different embedding dims

Commit:
Small code simplification

Commit:
Update docs in homology/simplicial.py following refactoring of check_point_clouds

Commit:
Revert to trivial slicing in transform_plot method of PlotterMixin

Needed to cover the case of list input

wreise commented on gtda/base.py:
```diff
@@ -139,7 +139,7 @@ def transform_plot(self, X, sample=0, **plot_params):
 
         """
-        Xt = self.transform(X[[sample]])
+        Xt = self.transform(X[sample:sample+1])
```
This is something that we use in other places as well. Does this become the accepted way of extracting one sample?

ulupo replied:
It was required here because the existing approach cannot cover lists. It could become the accepted way "just in case", but for now we use arrays everywhere other than in `homology/simplicial.py` so changing things is not urgent in my opinion.
